### PR TITLE
support: add lateness trait

### PR DESCRIPTION
**Pull request by andresilva**
Add a trait for reporting how late the current block is and implement it for BABE.

### CODE REVIEW

File: frame/babe/src/lib.rs [hunk 1]
```diff
@@ -443,7 +448,15 @@ impl<T: Trait> Module<T> {
...
 
-			CurrentSlot::put(digest.slot_number());
+			// the slot number of the current block being initialized
+			let current_slot = digest.slot_number();
+
+			// how many slots were skipped between current and last block
+			let lateness = current_slot.saturating_sub(CurrentSlot::get() + 1);
+			let lateness = T::BlockNumber::from(lateness as u32);
+
+			Lateness::<T>::put(lateness);
```
Comment (gavofyork) [inline frame/babe/src/lib.rs:458]: We're trying to cut down on extra DB I/O as it's pretty slow. Could we get this merged into `CurrentSlot`?

Comment (andresilva) [reply frame/babe/src/lib.rs:463]: This is always killed in `on_finalize` and should never reach the DB.

File: frame/babe/src/lib.rs [hunk 2]
```diff
@@ -498,6 +516,14 @@ impl<T: Trait> frame_support::traits::EstimateNextSessionRotation<T::BlockNumber
 	}
 }
 
+impl<T: Trait> frame_support::traits::Lateness<T::BlockNumber> for Module<T> {
+	fn lateness(&self) -> T::BlockNumber {
+		use sp_runtime::traits::Zero;
+
+		Self::lateness().unwrap_or(Zero::zero())
```
Comment (bkchr) [inline frame/babe/src/lib.rs:523]:
```suggestion
		Self::lateness().unwrap_or_default()
```

File: frame/babe/src/lib.rs [hunk 3]
```diff
@@ -152,6 +152,13 @@ decl_storage! {
 		/// Temporary value (cleared at block finalization) which is `Some`
 		/// if per-block initialization has already been called for current block.
 		Initialized get(fn initialized): Option<MaybeVrf>;
+
+		/// How late the current block is compared to its parent.
+		///
+		/// This entry is populated as part of block execution and is cleaned up
+		/// on block finalization. Querying this storage entry outside of block
+		/// execution context should always yield `None`.
+		Lateness get(fn lateness): Option<T::BlockNumber>;
```
Comment (bkchr) [inline frame/babe/src/lib.rs:161]: I don't understand why this is an `Option`.

If we remove the `Option` and call `kill()` and `get()` afterwards we will get `0`.

Comment (andresilva) [reply frame/babe/src/lib.rs]: Yeah, updated.

**Comment by bkchr** [comment]
You need to merge master to make the polkadot build green :)

**Pull request merged**